ghga_connector/core/upload.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2021 - 2023 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module dealing with intermediate upload path abstractions"""
+
+import math
+from pathlib import Path
+
+import crypt4gh.lib
+
+from ghga_connector.core import exceptions
+from ghga_connector.core.api_calls import Uploader
+from ghga_connector.core.client import async_client
+from ghga_connector.core.file_operations import Crypt4GHEncryptor
+from ghga_connector.core.message_display import AbstractMessageDisplay
+
+
+class ChunkedUploader:
+    """Handler class dealing with upload functionality"""
+
+    def __init__(
+        self, encryptor: Crypt4GHEncryptor, file_path: Path, uploader: Uploader
+    ) -> None:
+        self.encrypted_file_size = 0
+        self.encryptor = encryptor
+        self.file_id = uploader._file_id
+        self.input_path = file_path
+        self.part_size = uploader.part_size
+        self.unencrypted_file_size = file_path.stat().st_size
+        self.uploader = uploader
+
+    async def encrypt_and_upload(self):
+        """Delegate encryption and perform multipart upload"""
+
+        # compute encrypted_file_size
+        num_segments = math.ceil(self.unencrypted_file_size / crypt4gh.lib.SEGMENT_SIZE)
+        expected_encrypted_size = self.unencrypted_file_size + num_segments * 28
```

Comment:
I know this part was copied over from the DSK, but where does the magic 28 come from? It'd be nice to have that assigned to some descriptive name.

Reply:
This is now directly imported from Crypt4GH instead in 11ae731bdbd1ec1601b5b09409c3ee13938f3623